src/server/conn.rs

```diff
@@ -114,48 +130,145 @@ impl Conn {
-                    // we need to read more data for header
-                    break;
-                }
-
-                // we have already read whole header, parse it and begin to read payload.
-                let (msg_id, payload_len) = try!(rpc::decode_msg_header(self.header
-                    .bytes()));
-                self.last_msg_id = msg_id;
-                self.payload = Some(create_mem_buf(payload_len));
+    fn handshake<T, S>(&mut self,
+                       event_loop: &mut EventLoop<Server<T, S>>,
+                       bufs: &mut Vec<ConnData>)
+                       -> Result<()>
+        where T: RaftStoreRouter,
+              S: StoreAddrResolver
+    {
+        let mut data = match try!(self.read_one_message()) {
+            Some(data) => data,
+            None => return Ok(()),
+        };
+        if data.is_snapshot() {
+            self.conn_type = ConnType::Snapshot;
+
+            let mut snap_data = RaftSnapshotData::new();
+            try!(snap_data.merge_from_bytes(
+                data.msg.get_raft().get_message().get_snapshot().get_data()));
+            self.file_size = snap_data.get_file_size() as usize;
+            self.payload = Some(create_mem_buf(cmp::min(SNAPSHOT_PAYLOAD_BUF, self.file_size)));
+
+            let register_task = SnapTask::Register(self.token, data.msg.take_raft());
+            box_try!(self.snap_scheduler.schedule(register_task));
+
+            return self.read_snapshot(event_loop);
+        }
+        bufs.push(data);
+        self.conn_type = ConnType::Rpc;
+        self.read_rpc(event_loop, bufs)
+    }
+
+    fn read_snapshot<T, S>(&mut self, _: &mut EventLoop<Server<T, S>>) -> Result<()>
+        where T: RaftStoreRouter,
+              S: StoreAddrResolver
+    {
+        // TODO: limit rate
+        while try!(self.read_payload()) {
+            try!(self.handle_snapshot_payload());
+        }
+        Ok(())
+    }
+
+    fn handle_snapshot_payload(&mut self) -> Result<()> {
+        let mut payload = self.payload.take().unwrap();
+        let cap = payload.capacity();
+
+        let task = SnapTask::Write(self.token, payload.flip());
+        box_try!(self.snap_scheduler.schedule(task));
+
+        if self.read_size + cap == self.file_size {
+            // last chunk
+            box_try!(self.snap_scheduler.schedule(SnapTask::Close(self.token)));
+            if let Err(e) = self.sock.shutdown(Shutdown::Both) {
+                error!("shutdown connection error: {}", e);
             }
+        }
 
-            // payload here can't be None.
-            let mut payload = self.payload.take().unwrap();
-            try!(try_read_data(&mut self.sock, &mut payload));
-            if payload.remaining() > 0 {
-                // we need to read more data for payload
-                self.payload = Some(payload);
-                break;
+        payload = if self.read_size == self.file_size {
+            create_mem_buf(1)
```

Comment:
To break the while loop outside. Never mind, I inline this function and just break now.